### ISSUE TITLE
Migrate LightPcapNg to the forked repo

### ISSUE THREAD

**Issue by tigercosmos** [seladb/PcapPlusPlus#1348]
Currently we copy LightPcapNg code directly in `seladb/PcapPlusPlus `

The code lives at here:  [PcapPlusPlus](https://github.com/seladb/PcapPlusPlus/tree/1ff377c5de22caf9907fcc31f3f71ea138346155)/[3rdParty](https://github.com/seladb/PcapPlusPlus/tree/1ff377c5de22caf9907fcc31f3f71ea138346155/3rdParty)/[LightPcapNg](https://github.com/seladb/PcapPlusPlus/tree/1ff377c5de22caf9907fcc31f3f71ea138346155/3rdParty/LightPcapNg)

We now use the [current_commit.git](https://github.com/seladb/PcapPlusPlus/blob/1ff377c5de22caf9907fcc31f3f71ea138346155/3rdParty/LightPcapNg/current_commit.git) file to record the upstream git commit, but this way is difficult to sync the code from the upstream (basically, manual picking).

Therefore we created a forked LightPcapNg for esiast maintenance.

Here are the steps:
1. Migrate the current code to https://github.com/PcapPlusPlus/LightPcapNg (Manually picking and copying)
2. Remove the current copied code of LightPcapNg.
3. Use CMake `ExternalProject_Add` feature to download the source code from `PcapPlusPlus/LightPcapNg` when doing the first run of `cmake`. 
4. Try to push our change in `PcapPlusPlus/LightPcapNg` back to the upstream

**Comment by seladb** [comment]
@tigercosmos I don't think we should use git submodule. Unfortunately submodules are not very convenient because the user needs to remember to fetch them (a simple `git clone` won't fetch them by default).
I think we can keep what we do now, just sync the git commit against our fork instead of the upstream repo

**Comment by tigercosmos** [comment]
@seladb  I think we can just write in the document to use `git clone --recurse-submodules`?

> I think we can keep what we do now, just sync the git commit against our fork instead of the upstream repo

Do you mean to copy the code from the forked repo? I think it doesn't make the maintaining sync issue easier.

**Comment by seladb** [comment]
> @seladb I think we can just write in the document to use `git clone --recurse-submodules`?
> 
> > I think we can keep what we do now, just sync the git commit against our fork instead of the upstream repo
> 
> Do you mean to copy the code from the forked repo? I think it doesn't make the maintaining sync issue easier.

yes... that's the reason I copied the code in the first place. Enforcing doing `git clone --recurse-submodules` is pretty annoying. Please notice we do that for all 3rd party libs. Since it's a forked repo that we don't change a lot, it shouldn't be very hard to maintain our copied code the same as the fork...

**Comment by tigercosmos** [comment]
I disagree with duplicating the code. If we opt to increment the version in the forked repository, it would result in copying all the code into PCPP. This approach seems inefficient. Instead, why not simply increment the commit in the git submodule?

Kindly permit me to invite @clementperon and @egecetin to share their opinions on this matter as well.

**Comment by egecetin** [comment]
To be honest submodule is not a problem for experienced developers they can easily figure out when they even configuring project with CMake and get path does not exist or similar error. But for most inexperienced people it is quite hard to figure out they missed the submodule init (Why does the git clone default do not fetch submodules, they are part of code, so I think also which is a bit annoying). If the concern is only troubles caused by missing the submodule initialization maybe we can add a warning line to CMake to remind people to call `git submodule update --init --recursive` (since they already cloned this command is better) if the path does not exist?

**Comment by tigercosmos** [comment]
@seladb  how about this way, we don't copy the forked code into PCPP nor use the git submodule. 

When the user runs `cmake` for the first time in PCPP, make will fetch and get the code of LightPcapNg from the forked repo.

For example:
```
include(ExternalProject)

# Define the external project
ExternalProject_Add(
    MyProjectExternal    # Arbitrary name for the project
    PREFIX ${CMAKE_BINARY_DIR}/MyProject    # Directory where the project will be downloaded and built
    GIT_REPOSITORY https://github.com/username/repo.git    # URL of the Git repository
    GIT_TAG master    # Branch, tag, or commit hash to checkout
    UPDATE_DISCONNECTED 1    # Don't update the project during the build
    # You can also specify other options such as CONFIGURE_COMMAND, BUILD_COMMAND, INSTALL_COMMAND, etc.
)

# Add dependencies of your main project to the external project
add_dependencies(MyMainProject MyProjectExternal)
```

**Comment by seladb** [comment]
> ExternalProject_Add

Hmm I'm open to this idea! Can you please create a POC of this to see how it works?

**Comment by tigercosmos** [comment]
@seladb  sure, revised the issue content.

**Comment by Dimi1010** [comment]
With regards to this, do we have an issue with exporting `light_pcapng.lib` static alongside the `pcapplusplus` install, as a vendored library? Currently the library is inlined into `Pcap++.lib` due to the vendored version redefining it as a `object` library but the fork defines the targets as static library.

This is further complicated by `Pcap++` also linking to `light_pcapng` library if `zstd` flag is set to `true`.

**Comment by seladb** [comment]
> With regards to this, do we have an issue with exporting `light_pcapng.lib` static alongside the `pcapplusplus` install, as a vendored library? Currently the library is inlined into `Pcap++.lib` due to the vendored version redefining it as a `object` library but the fork defines the targets as static library.
> 
> This is further complicated by `Pcap++` also linking to `light_pcapng` library if `zstd` flag is set to `true`.

@Dimi1010 I think we shouldn't export `light_pcapng.lib` as a separate library but embed it inside `Pcap++.lib` like we do today. I'm not sure what's the right way to do it and if CMake provides such an option 🤔

**Comment by Dimi1010** [comment]
> [@Dimi1010](https://github.com/Dimi1010) I think we shouldn't export `light_pcapng.lib` as a separate library but embed it inside `Pcap++.lib` like we do today. I'm not sure what's the right way to do it and if CMake provides such an option 🤔

@seladb Is there a specific reason for that or just because it looks nice?

**Comment by seladb** [comment]
> [@seladb](https://github.com/seladb) Is there a specific reason for that or just because it looks nice?

It's because we don't want to expose our "internal implementation" - we should only expose 3 libs. The fact that we use 3rd-party libraries shouldn't matter to our users - it should all be included in one of the 3 libs

**Comment by Dimi1010** [comment]
I think it should be fine to vendor the .lib files provided the cmake buildsystem does the setup seamlessly for the users.

It also simplifies the build system.

Currently, the fork provides `static` target to `light_pcapng`. To inline the object files, the target must be of type `object`, so that would mean significant changes to the fork's build system.

I did a test in the FetchContent PR to have the `light_pcapng` linked statically. The install artifacts end up thus:

```
root
- bin
- include
- lib
-- cmake
-- vendor
--- light_pcapng.lib
-- Pcap++.lib
-- Packet++.lib
-- Common++.lib
```

No headers from `light_pcapng` are exported publicly, and the `Pcap++` seamlessly sets up a private dependency on the `light_pcapng` target.

**Comment by seladb** [comment]
@Dimi1010 I created a simple POC project with a lib that uses a third-party lib, and managed to get it working the way we want. We need to build LightPcapNg as `OBJECT` and then add its `.o` files to `libPcap++.a`. I'm trying it out now on [my branch](https://github.com/seladb/PcapPlusPlus/pull/1792). I also added an option in the forked repo to choose the build type: https://github.com/PcapPlusPlus/LightPcapNg/blob/copy-from-pcapplusplus/CMakeLists.txt

I'll let you know if I managed to get it working.

**Comment by Dimi1010** [comment]
@seladb That should work, although I recommend having 3 separate targets for the different library types in the fork, rather than changing the target type by variable.

**Comment by seladb** [comment]
Done in https://github.com/seladb/PcapPlusPlus/pull/1792
